### PR TITLE
[flutter_local_notifications] Implement options to hide or crop attachments in the thumbnail on iOS

### PR DESCRIPTION
**Pull request by noinskit**
This PR implements access to two platform-specific features for customizing attachments to notifications (c.f. https://developer.apple.com/documentation/usernotifications/unnotificationattachment?language=objc). Without them it's very hard to use image attachments at all, because by default the first attachment is used both for a tiny thumbnail (smaller than app logo) and a potentially full screen image on an expanded notification.

### CODE REVIEW

**Comment by MaikuB** [comment]
Can you merge the latest changes from master to your branch/fork? This would fix the Kotlin plugin issue and allow seeing if tests are failing or not

File: flutter_local_notifications/lib/src/platform_specifics/darwin/notification_attachment.dart
Comment (MaikuB) [inline flutter_local_notifications/lib/src/platform_specifics/darwin/notification_attachment.dart:30]: - Can you add public API docs for these? This should've picked up by the linter
- Are you able to add an example that makes use of this?

**Review by MaikuB — CHANGES_REQUESTED** [review]

File: flutter_local_notifications/lib/src/platform_specifics/darwin/notification_attachment.dart
Comment (noinskit) [reply flutter_local_notifications/lib/src/platform_specifics/darwin/notification_attachment.dart:31]: Both done!

**Comment by MaikuB** [comment]
Good call out as I believe the API being used is also supported on macOS and my stance is that macOS support should be added for situations like this. On that note, are you able to make the macOS changes too or would you need support for that?

**Comment by noinskit** [comment]
> Good call out as I believe the API being used is also supported on macOS and my stance is that macOS support should be added for situations like this. On that note, are you able to make the macOS changes too or would you need support for that?

I wanted to just implement, but I need help after all. I started the "example" app on my Mac (Ventura 13.0) and all the buttons silently do nothing, including `Show plain notification with payload` and `Request permission`. I know that e.g. Chrome sends me system notifications on the same system. Nothing in IntelliJ or XCode logs.

**Comment by noinskit** [comment]
I'd add that I've found https://github.com/MaikuB/flutter_local_notifications/issues/1610 and tried a bunch of things to see a system dialog to ask me for permissions, incl. deleting `Derived Data`, `killall NotificationCenter` and some other things, with no luck. I cannot find it anywhere in `System Settings` either.

**Comment by MaikuB** [comment]
Figured it out and will submit a PR to fix this soon. It looks like requesting critical alert permissions has prevented the permissions dialog appearing on macOS but was fine for me on iOS. Not sure why the behaviour is inconsistent

**Comment by MaikuB** [comment]
If you update your branch/fork from the master branch then this should be in the change since https://github.com/MaikuB/flutter_local_notifications/pull/1809 is merged in

**Comment by noinskit** [comment]
> If you update your branch/fork from the master branch then this should be in the change since #1809 is merged in

Thanks, I got the permission dialog now and I can trigger notifications! I'm looking at the thumbnail options on macOS now...

**Comment by noinskit** [comment]
I pushed a version with macOS support. BUT it's not really working for me on my laptop, even though as far as I can tell it's being passed correctly (I looked at debug prints of the constructed `options` hashtable). I'm thinking that maybe these options aren't really supported on macOS yet. If that's the case, passing them along might still be correct to do in the plugin and some future macOS version might learn to respect them. That said, I'd appreciate if you took a look if something else could be wrong with the code!

**Comment by MaikuB** [comment]
No worries, I'll take a look and see if I spot anything

**Comment by MaikuB** [comment]
Whilst I take a look at this, could you add public API docs? The linter rules should highlight the ones that are missing

**Comment by MaikuB** [comment]
Took a look and ran into same problem. I don't think there's anything wrong with what was implemented. The only thing that came to mind that didn't change the result is that the official docs for [UNNotificationAttachmentOptionsThumbnailClippingRectKey](https://developer.apple.com/documentation/usernotifications/unnotificationattachmentoptionsthumbnailclippingrectkey) say to use [this](https://developer.apple.com/documentation/coregraphics/1455760-cgrectcreatedictionaryrepresenta) to create the dictionary representation. Although it seems to have not made a difference, perhaps the code should change to use that function?

**Comment by noinskit** [comment]
Sure, I made both changes! Still no impact on how the notification is shown on macOS, though :(

**Comment by noinskit** [comment]
Or maybe not, looks like tests are now failing on:
`error: 'CGRectCreateDictionaryRepresentation' has been replaced by property 'CGRect.dictionaryRepresentation'`

I'm almost sure it worked for me locally; I'm guessing it's about compiler settings.

**Comment by MaikuB** [comment]
Yeah it worked locally for me too. Perhaps the docs are out of date given the message. Looks like that particular change would need to be reverted as it sounds like what you did is more correct.

On a different note, in case you were squashing commits out of concern that it would affect the commit history when merged then don't worry about that GitHub provides the ability to do the maintainer's end before merging and it's what I select as the option too

**Comment by noinskit** [comment]
Sure, it's back to `dictionaryRepresentation`.

File: flutter_local_notifications/ios/Classes/FlutterLocalNotificationsPlugin.m
Comment (MaikuB) [inline flutter_local_notifications/ios/Classes/FlutterLocalNotificationsPlugin.m:935]: for consistency and to avoid any potential issues down the line given what we saw on macOS, can this go back to using dictionaryRepresentation like it did before?

Comment (noinskit) [reply flutter_local_notifications/ios/Classes/FlutterLocalNotificationsPlugin.m:935]: I'm not sure if I understand... This part was always like that. The `dictionaryRepresentation` method on `CGRect` is there in Swift, but in ObjC this is it.

Comment (MaikuB) [reply flutter_local_notifications/ios/Classes/FlutterLocalNotificationsPlugin.m:935]: Apologies misremembered and turns out it's not even available in Objective-C

**Review by MaikuB — APPROVED** [review]

**Pull request merged**
